controller/eks-cluster-config-handler.go

```diff
@@ -1020,21 +1062,47 @@ func (h *Handler) updateUpstreamClusterState(upstreamSpec *eksv1.EKSClusterConfi
 		// happen together
 
 		ng := ngs[aws.StringValue(upstreamNg.NodegroupName)]
-		if ng.Version != nil {
-			if aws.StringValue(upstreamNg.Version) != desiredNgVersions[aws.StringValue(ng.NodegroupName)] {
-				attemptUpdatingNodegroups = true
-				_, err := eksService.UpdateNodegroupVersion(
-					&eks.UpdateNodegroupVersionInput{
-						NodegroupName: aws.String(aws.StringValue(ng.NodegroupName)),
-						ClusterName:   aws.String(config.Spec.DisplayName),
-						Version:       aws.String(desiredNgVersions[aws.StringValue(ng.NodegroupName)]),
-					},
-				)
+		ngVersionInput := &eks.UpdateNodegroupVersionInput{
+			NodegroupName: aws.String(aws.StringValue(ng.NodegroupName)),
+			ClusterName:   aws.String(config.Spec.DisplayName),
+		}
+
+		if upstreamNg.LaunchTemplate != nil {
+			upstreamTemplateVersion := aws.Int64Value(upstreamNg.LaunchTemplate.Version)
+			var err error
+			lt := ng.LaunchTemplate
+
+			if lt == nil && config.Status.LaunchTemplateID == aws.StringValue(upstreamNg.LaunchTemplate.ID) {
+				// In this case, Rancher is managing the launch template, so we check to see if we need a new version.
+				lt, err = newLaunchTemplateVersionIfNeeded(config, upstreamNg, ng, ec2Service)
 				if err != nil {
 					return config, err
 				}
-				continue
 			}
+
+			if lt != nil && (aws.StringValue(lt.ID) != aws.StringValue(upstreamNg.LaunchTemplate.ID) || aws.Int64Value(lt.Version) != upstreamTemplateVersion) {
+				attemptUpdatingNodegroups = true
+				ngVersionInput.LaunchTemplate = &eks.LaunchTemplateSpecification{
+					Id:      lt.ID,
+					Version: aws.String(strconv.FormatInt(*lt.Version, 10)),
+				}
+				if upstreamTemplateVersion > 0 {
+					templateVersionsToDelete = append(templateVersionsToDelete, aws.String(strconv.FormatInt(upstreamTemplateVersion, 10)))
+				}
+			}
+		} else if ng.Version != nil {
```

Comment:
Doesn't version here refer to kubernetes version? Don't we want to check this regardless of whether a template is used?

Reply:
Took a look at the sdk code. So is the `UpdateNodegroupVersionInput` struct and `UpdateNodegroupVersion` function are now used to upgrade launch template versions or kubernetes version? Why are these operations mutually exclusive? I thought launch templates were specifically concerned with ec2 intsance configuration?

Reply:
I took away the else statement, so `LaunchTemplate` and `Version` can be updated in a single call.

Reply:
The problem wasn't whether it did it in a single call, it just looked like it would not upgrade thee kubernetes version if there was a launch template.

Reply:
This looks good though, will resolve pending other question about this logic.